diff --git a/sysdeps/aarch64/memcmp.S b/sysdeps/aarch64/memcmp.S
index 827f54f99e..0159102045 100644
--- a/sysdeps/aarch64/memcmp.S
+++ b/sysdeps/aarch64/memcmp.S
@@ -41,7 +41,11 @@
 #define tmp1		x7
 #define tmp2		x8
 
-ENTRY_ALIGN (memcmp, 6)
+#ifndef MEMCMP
+# define MEMCMP memcmp
+#endif
+
+ENTRY_ALIGN (MEMCMP, 6)
 	DELOUSE (0)
 	DELOUSE (1)
 	DELOUSE (2)
@@ -174,7 +178,7 @@ L(ret_0):
 	mov	result, 0
 	ret
 
-END (memcmp)
+END (MEMCMP)
 #undef bcmp
-weak_alias (memcmp, bcmp)
-libc_hidden_builtin_def (memcmp)
+weak_alias (MEMCMP, bcmp)
+libc_hidden_builtin_def (MEMCMP)
diff --git a/sysdeps/aarch64/multiarch/Makefile b/sysdeps/aarch64/multiarch/Makefile
index fc95c0f62a..64d227803c 100644
--- a/sysdeps/aarch64/multiarch/Makefile
+++ b/sysdeps/aarch64/multiarch/Makefile
@@ -2,6 +2,7 @@ ifeq ($(subdir),string)
 sysdep_routines += memcpy_generic memcpy_thunderx memcpy_thunderx2 memcpy_kunpeng950 \
 		   memcpy_falkor memmove_falkor \
 		   memset_generic memset_falkor memset_emag memset_kunpeng memset_kunpeng950 \
+		   memcmp_generic memcmp_kunpeng950 \
 		   memchr_generic memchr_nosimd \
 		   strlen_generic strlen_asimd
 endif
diff --git a/sysdeps/aarch64/multiarch/memcmp.c b/sysdeps/aarch64/multiarch/memcmp.c
new file mode 100644
index 0000000000..fd69d8c694
--- /dev/null
+++ b/sysdeps/aarch64/multiarch/memcmp.c
@@ -0,0 +1,41 @@
+/* Multiple versions of memcmp. AARCH64 version.
+   Copyright (C) 2018-2020 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+/* Define multiple versions only for the definition in libc.  */
+
+#if IS_IN (libc)
+/* Redefine memcmp so that the compiler won't complain about the type
+   mismatch with the IFUNC selector in strong_alias, below.  */
+# undef memcmp
+# define memcmp __redirect_memcmp
+# include <string.h>
+# include <init-arch.h>
+
+extern __typeof (__redirect_memcmp) __memcmp;
+
+extern __typeof (__redirect_memcmp) __memcmp_generic attribute_hidden;
+extern __typeof (__redirect_memcmp) __memcmp_kunpeng950 attribute_hidden;
+
+libc_ifunc (__memcmp,
+	    ((IS_KUNPENG950(midr)
+	       ? __memcmp_kunpeng950
+	       : __memcmp_generic)));
+
+# undef memcmp
+strong_alias (__memcmp, memcmp);
+#endif
diff --git a/sysdeps/aarch64/multiarch/memcmp_generic.S b/sysdeps/aarch64/multiarch/memcmp_generic.S
new file mode 100644
index 0000000000..bf3c263c52
--- /dev/null
+++ b/sysdeps/aarch64/multiarch/memcmp_generic.S
@@ -0,0 +1,33 @@
+/* Memcmp for aarch64, default version for internal use.
+   Copyright (C) 2018-2020 Free Software Foundation, Inc.
+
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library.  If not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#if IS_IN (libc)
+# define MEMCMP __memcmp_generic
+
+/* Do not hide the generic version of memcmp, we use it internally.  */
+# undef libc_hidden_builtin_def
+# define libc_hidden_builtin_def(name)
+
+/* Add a hidden definition for use within libc.so.  */
+# ifdef SHARED
+	.globl __GI_memcmp; __GI_memcmp = __memcmp_generic
+# endif
+#endif
+
+# include "../memcmp.S"
diff --git a/sysdeps/aarch64/multiarch/memcmp_kunpeng950.S b/sysdeps/aarch64/multiarch/memcmp_kunpeng950.S
new file mode 100644
index 0000000000..354f2af4de
--- /dev/null
+++ b/sysdeps/aarch64/multiarch/memcmp_kunpeng950.S
@@ -0,0 +1,109 @@
+/* Optimized memcmp for Huawei Kunpeng processor.
+   Copyright (c) 2025 Huawei Technologies Co., Ltd.
+   Copyright (C) 2012-2020 Free Software Foundation, Inc.
+
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library.  If not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include <sysdep.h>
+
+#if IS_IN (libc)
+# define MEMCMP __memcmp_kunpeng950
+
+/* Assumptions:
+ *
+ * ARMv8.2-a, AArch64, unaligned accesses
+ *
+ */
+
+.arch armv8.2-a+sve
+
+ENTRY_ALIGN (MEMCMP, 6)
+	cbz	x2, .Lequal
+	mov	x3, #0
+	index	z5.b, #0, #1
+
+1:	whilelo	p0.b, x3, x2
+	b.none	.Lequal
+
+	addvl	x7, x3, #1
+	addvl	x8, x3, #2
+	addvl	x9, x3, #3
+
+
+	ld1b	z0.b, p0/z, [x0, x3]
+	ld1b	z1.b, p0/z, [x1, x3]
+	cmpne	p1.b, p0/z, z0.b, z1.b
+	b.any	.Lmm0
+
+	whilelo	p2.b, x7, x2
+	b.none	.Lequal
+	ld1b	z2.b, p2/z, [x0, x7]
+	ld1b	z3.b, p2/z, [x1, x7]
+	cmpne	p3.b, p2/z, z2.b, z3.b
+	b.any	.Lmm1
+
+	whilelo	p4.b, x8, x2
+	b.none	.Lequal
+	ld1b	z6.b, p4/z, [x0, x8]
+	ld1b	z7.b, p4/z, [x1, x8]
+	cmpne	p5.b, p4/z, z6.b, z7.b
+	b.any	.Lmm2
+
+	whilelo	p6.b, x9, x2
+	b.none	.Lequal
+	ld1b	z16.b, p6/z, [x0, x9]
+	ld1b	z17.b, p6/z, [x1, x9]
+	cmpne	p7.b, p6/z, z16.b, z17.b
+	b.any	.Lmm3
+
+	addvl	x3, x3, #4
+	b	1b
+
+.Lmm0:
+	brkb	p2.b, p0/z, p1.b
+	lasta	w0, p2, z0.b
+	lasta	w1, p2, z1.b
+	sub	x0, x0, x1
+	ret
+
+.Lmm1:
+	brkb	p0.b, p2/z, p3.b
+	lasta	w0, p0, z2.b
+	lasta	w1, p0, z3.b
+	sub	x0, x0, x1
+	ret
+
+.Lmm2:
+	brkb	p6.b, p4/z, p5.b
+	lasta	w0, p6, z6.b
+	lasta	w1, p6, z7.b
+	sub	x0, x0, x1
+	ret
+
+.Lmm3:
+	brkb	p4.b, p6/z, p7.b
+	lasta	w0, p4, z16.b
+	lasta	w1, p4, z17.b
+	sub	x0, x0, x1
+	ret
+
+.Lequal:
+	mov	x0, #0
+	ret
+END (MEMCMP)
+libc_hidden_builtin_def (MEMCMP)
+#endif
